### DISCUSSION TITLE
Instrument programming model

### DISCUSSION THREAD

**Discussion by vatgh** [pymeasure/pymeasure#402] [Ideas]
Before contributing with some drivers for instruments I wanted to clarify the take of this project on which model to use. 
I could not find any reference to IVI or SCPI as a reference model in the documentation. Therefore, if I implement for example an oscilloscope, which model should I use in order to make the drivers for this oscilloscope work the same way as for all other oscilloscopes?
IVI addresses exactly this question (see http://www.ivifoundation.org/downloads/IVI_GSG_v_1.0.pdf in the introduction) with different classes for common instruments (oscilloscopes, multimeters, ...)
SCPI has only very basic common, defined comments, as far as I can tell (http://www.ivifoundation.org/docs/scpi-99.pdf)
Alternatively, one could use the structure that the commands of the instruments provide. But I think there is not necessarily large overlap from one instrument to the other. Small overlap would result in the need to modify code when swapping instruments which would be annoying.

I would prefer IVI as it addresses exactly this question and I do not see a need to reinvent the wheel. I would not implement all IVI functions for each instrument but only some basic ones to start with.

Btw I know that there is Python IVI (https://github.com/python-ivi/python-ivi) which has already a lot of driver implemented. Maybe one could make use of this.

**Comment by bilderbuchi** [comment]
You might also take a look at InstrumentKit, which has [abstract instruments](https://github.com/Galvant/InstrumentKit/tree/master/instruments/abstract_instruments) like you suggest.

**Comment by cjermain** [comment]
@vatgh, thanks for your thoughts on IVI vs. SCPI. We have typically focused on SCPI commands.

To give you more context, the PyMeasure project mainly focuses on the experiment execution and graphical user interfaces -- to make experiments easy to run and manage. For convenience, we have a fairly mature set of instrument classes. However, as @bilderbuchi suggested there are other projects that would be a better fit for deeper contributions on the instrument models. There has been some attempt to consolidate (#53), since many people are re-inventing the wheel on this problem. I think you are definitely on the right track with your approach.

**Comment by vatgh** [comment]
Thanks for the replies. I am aware of the long discussion #53 but I did not see any result which I guess means that there are no "plug&play" interfaces in between the libraries/projects. Correct?
Now we would like to start using PyMeasure but for this we have to implement our instruments. And in order to do this properly I would have to start with a model. SCPI does not really provide this. 
Alternatively, I could contribute to an effort to make existing drivers from InstrumentKit (I have not done anything with this) or Python-IVI (which I looked into a little bit) usable in PyMeasure but also for this it would be good to have some input from the lead developers of the latter. In particular because my Python coding skills are not at a particularly good level which makes it difficult for me to judge the best way to implement such a "mid-layer".

**Comment by bilderbuchi** [comment]
At the end of that discussion there are some links to further (hopefully more structured) discussion in the labpy organisation. In one issue we are currently trying to catalog the python instrument driver packages out there - help is welcome!

Making IK drivers available in pymeasure would be cool (there is an issue about this in the IK issue tracker), but I'm not aware of any progress on that front.

**Comment by bilderbuchi** [comment]
See #94 for the issue about compatibility with IK

**Comment by bilderbuchi** [comment]
This issue has more of a discussion/usage question character, so I will move it to our new discussion forum, which is a better fit than the issue tracker.
